text/0000-conservative-impl-trait.md

```diff
@@ -0,0 +1,553 @@
+introduction of abstract types limited to the return type of inherent methods
+and free functions. While we still need to resolve some of the core questions
+about what an "abstract type" means even in these cases, we avoid some of the
+complexities that come along with allowing the feature in other locations or
+with other extensions.
+
+## Syntax
+
+Let's start with the bikeshed: The proposed syntax is `@Trait` in return type
+position, composing like trait objects to forms like `@(Foo+Send+'a)`.
+
+The reason for choosing a sigil is ergonomics: Whatever the exact final feature
+will be capable of, you'd want it to be as easy to read/write as trait objects,
+or else the more performant and idiomatic option would be the more verbose one,
+and thus probably less used.
+
+The argument can be made this decreases the google-ability of Rust syntax (and
+this doesn't even talk about the _old_ `@T` pointer semantic the internet is
+still littered with), but this would be somewhat mitigated by the feature being
+supposedly used commonly once it lands, and can be explained in the docs as
+being short for `abstract` or `anonym`. And in any case, it's a problem we
+already suffer with `&T` and `&mut T`.
+
+If there are good reasons against `@`, there is also the choice of `~`.
+All points from above still apply, except `~` is a bit rarer in language
+syntaxes in general, and depending on keyboard layout somewhat harder to reach.
+
+Finally, if there is a huge incentive _against_ new (old?) sigils in the language,
+there is also the option of using keyword-based syntax like `impl Trait` or
+`abstract Trait`, but this would add a verbosity overhead for a feature
+that will be used somewhat commonly.
+
+## Semantics
+
+The core semantics of the feature is described below.
+
+Note that the sections after this one go into more detail on some of the design
+decisions, and that **it is likely for many of the mentioned limitations to be
+lifted at some point in the future**. For clarity, we'll separately categories the *core
+semantics* of the feature (aspects that would stay unchanged with future extensions)
+and the *initial limitations* (which are likely to be lifted later).
+
+**Core semantics**:
+
+- If a function returns `@Trait`, its body can return values of any type that
+  implements `Trait`, but all return values need to be of the same type.
+
+- As far as the typesystem and the compiler is concerned, the return type
+  outside of the function would not be a entirely "new" type, nor would it be a
+  simple type alias. Rather, its semantics would be very similar to that of
+  _generic type paramters_ inside a function, with small differences caused by
+  being an _output_ rather than an _input_ of the function.
+
+  - The type would be known to implement the specified traits.
+  - The type would not be known to implement any other trait, with
+    the exception of OIBITS (aka "auto traits") and default traits like `Sized`.
```

Comment:
What happens when an OIBIT is added, maybe even in a non-`std` crate?
The thing you're proposing has been considered and it might work if `Send` and `Sync` were the only OIBITs ever. But that's wrong, even right now.